modules/social_features/social_event/modules/social_event_managers/src/Plugin/Action/SocialEventManagersSendEmail.php

```diff
@@ -144,4 +248,105 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
     return parent::buildConfigurationForm($form, $form_state);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function setContext(array &$context): void {
+    parent::setContext($context);
+
+    // @todo Rely on something more solid then dynamic data for batch.
+    // We don't want to run this code if batch was started.
+    // "prepopulated" key is adding exactly on batch start.
+    if (isset($context['prepopulated'])) {
+      return;
+    }
+
+    if (!isset($context['list'], $context['bulk_form_keys'])) {
+      // Probably, the method is executed on batch processing.
+      return;
+    }
+
+    // Filter enrollees that have disabled bulk mailing based on their profile
+    // settings.
+    // We need to remove the selected users from temporary data and update
+    // form data passed through the further forms.
+    if ($select_all = empty($context['list'])) {
+      $selected_options = $context['bulk_form_keys'];
+    }
+    else {
+      $selected_options = array_keys($context['list']);
+    }
+
+    if (!$selected_options) {
+      return;
+    }
+
+    $origin = [...$selected_options];
+
+    // Go through each enrollment and check if a user doesn't have
+    // a disabled bulk mailing.
+    foreach ($selected_options as $key => $name) {
+      $item = (array) $this->getListItem($name);
+      if (!in_array('event_enrollment', $item)) {
+        continue;
+      }
+
+      // First element is the enrollment ID.
+      $eid = $item[0] ?? 0;
+      $enrollment = EventEnrollment::load($eid);
+      if (!$enrollment) {
+        continue;
+      }
+
+      $account = $enrollment->getAccountEntity();
+
+      // Check user frequency settings for event bulk mailing.
+      // If a user has disabled mailing, we remove enrollment from
+      // the selected list.
+      // Only authenticated users have frequency settings.
+      if ($account && $this->isUnsubscribedFromEmails($account)) {
+        unset($selected_options[$key]);
+      }
+    }
```

Comment:
I don't see there any possible optimizations. Could you maybe put suggestions here?

Reply:
```php
    $event_enrollment_ids = [];
    foreach ($selected_options as $key => $name) {
      $item = (array) $this->getListItem($name);
      if (!in_array('event_enrollment', $item)) {
        continue;
      }

      // First element is the enrollment ID.
      if (isset($item[0])) {
        $event_enrollment_ids[$key] = $item[0];
      }
    }
    $event_enrollments = EventEnrollments::loadMultiple($event_enrollment_ids);
    $account_ids = [];
    foreach ($event_enrollment_ids as $key => $eid) {
      if (!isset($event_enrollments[$eid])) {
        continue;
      }
      $account_ids[$key] = $event_enrollment->getAccount();
    }
    $accounts = User::loadMultiple($account_ids);
    foreach ($account_ids as $key => $uid) {
      if (!isset($accounts[$uid])) {
        continue;
      }
      if ($this->unsubscribedFromEmails($accounts[$uid])) {
        unset($selected_options[$key]);
      }
    }
```

Uses 2 loops to turn loading individual event enrollments and users into a single query per entity type. The `unsubscribedFromEmails` unfortunately can't be changed to be efficient in its current form.

There is one trade-off. SInce we're working with bulk operations. In the form as initially written, a group with 1000 users (as far as I'm aware, those exist) where an e-mail would be sent to every user, would make 3000 database queries in this function alone. After my proposal this would be only 1002 database queries (a 66% improvement). With the trade-off that we try to hold 1000 event enrollments and 1000 users in memory at the same time, which likely causes different issues.

Similar to update hooks we likely want to implement some form of batching and load them in batches of ~50 entities. That would cause 1040 database queries to be performed, which would still be a 65,3% improvement, but without the memory issues. 

Ideally since we know this is going to be run with large selections of users, we'd also fix our `getBulkEmailUserSetting` in a way that we can actually do so for large numbers of users at a time. Given that it's a custom table, this should be doable.

Reply:
Thanks, I added (and tested) your suggestion.